### PR TITLE
LCOE calculations with parasitic load costs in numerator, Energy sales in denominator

### LINKED ISSUE #546: Fix Accounting for Inverter Operating Losses in LCOE/Gen
Currently, inverter operating losses (day and nighttime) are subtracted off of `gen`, which is the energy used in the numerator of the LCOE. However, clarification from a solar/wind developer indicates that the energy required for inverter operating losses should actually be counted as an operating cost, and that energy should NOT be subtracted from `gen`. The operating cost can be calculated either at the PPA price (if no retail rate is specified), OR at the utility rate. (We'll still need to figure out what to do for behind-the-meter systems)

From the developer: "For the cash flow models she uses during the development phase to determine a PPA price to use in proposals for solar projects, she gets a single annual kWh estimate from her engineering team to account for total inverter consumption (day and nighttime), and then applies either the PPA price or a retail rate to estimate the operating cost. Because the kWh value is relatively small for solar projects, some utilities don't bother billing at the retail rate and treat it as negative generation, but other utilities do charge a retail rate. Retail rates are typically commercial rates with an energy charge and demand charge. On the wind side, site consumption is a larger fraction of the total system output, so this value has a bigger impact on the cash flow."

### PR DESCRIPTION
**Pull request by mjprilliman**
-Updated LCOE calculation in all financial models
-Parasitic load operating cost based on PPA price or retail rate for FOM
-Parasitic load operating cost based on hourly prices for BTM
-Variable generation O&M calculated based on energy of value (FOM-sales, BTM-to grid or to load) not net energy
-LCOS updated with same LCOE methods for calculation of cost of grid charging
-Closes #546 outside of larger powerflow changes

### CODE REVIEW

**Comment by cpaulgilman** [comment]
@mjprilliman What branch of SAM is best to test this with?

**Comment by mjprilliman** [comment]
> @mjprilliman What branch of SAM is best to test this with?

Identical branch name in SAM. https://github.com/NREL/SAM/pull/749

**Review by cpaulgilman — COMMENTED** [review]
I'm not sure how to evaluate these changes to the LCOE, but it seems to be working as intended.

I added electricity purchases to cash flow for Detailed PV / PPA configs. Not sure if it should also be added for other technologies.

SAM crashes for CSP systems: 

1. Create CSP / Single Owner case.
2. Go to Electricity Purchases page
3. Choose **Retail electricity rate(s) option**

**Comment by mjprilliman** [comment]
> I'm not sure how to evaluate these changes to the LCOE, but it seems to be working as intended.
> 
> I added electricity purchases to cash flow for Detailed PV / PPA configs. Not sure if it should also be added for other technologies.
> 
> SAM crashes for CSP systems:
> 
> 1. Create CSP / Single Owner case.
> 2. Go to Electricity Purchases page
> 3. Choose **Retail electricity rate(s) option**

Either through code merges or a mistake on my part en_electricity_rates was calculated in the equations and updated in the callbacks, which was causing intermittent crashing for every configuration. Hoping the most recent push will fix the crashing, running the test script now.

**Comment by cpaulgilman** [comment]
If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate. 

As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.

I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.

File: ssc/cmod_equpartflip.cpp [hunk 1]
```diff
@@ -915,6 +919,9 @@ class cm_equpartflip : public compute_module
 	{
 		int i = 0;
 
+        if (is_assigned("en_electricity_rates") && as_number("en_electricity_rates") == 0 && as_number("ppa_soln_mode") == 0)
+            throw exec_error("singleowner", "PPA price from which to calculate parasitic load costs is not specified. Check inputs for Revenue and Electricity Purchases.");
```
Comment (brtietz) [inline ssc/cmod_equpartflip.cpp:923]: Please change the cmod name in the error message

File: ssc/cmod_cashloan.cpp [hunk 2]
```diff
@@ -990,7 +1104,7 @@ class cm_cashloan : public compute_module
 			lcoe_real /= npv_energy_real;
 		}
 
-		double npv_energy_nom = npv( CF_energy_net, nyears, nom_discount_rate );
+		double npv_energy_nom = npv( CF_energy_sales, nyears, nom_discount_rate ) + npv(CF_parasitic_cost, nyears, nom_discount_rate);
```
Comment (brtietz) [inline ssc/cmod_cashloan.cpp:1107]: I thought CF_parasitic_cost went in the numerator. This is the energy term in the denominator of the nominal LCOE.

File: ssc/cmod_host_developer.cpp [hunk 3]
```diff
@@ -2657,16 +2707,55 @@ class cm_host_developer : public compute_module
...
+	double host_npv_energy_real = npv(CF_energy_sales, nyears, host_disc_real);
+	double host_lcoe_real = -(cf.at(CF_after_tax_net_equity_cost_flow, 0) + npv(CF_after_tax_net_equity_cost_flow, nyears, host_nom_discount_rate) + npv(CF_parasitic_cost, nyears, nom_discount_rate)) * 100;
 	if (host_npv_energy_real == 0.0)
 		host_lcoe_real = std::numeric_limits<double>::quiet_NaN();
 	else
 		host_lcoe_real /= host_npv_energy_real;
 
-	double host_npv_energy_nom = npv(CF_energy_net, nyears, host_nom_discount_rate);
+	double host_npv_energy_nom = npv(CF_energy_sales, nyears, host_nom_discount_rate);
 	double host_lcoe_nom = -(cf.at(CF_after_tax_net_equity_cost_flow, 0) + npv(CF_after_tax_net_equity_cost_flow, nyears, host_nom_discount_rate)) * 100;
```
Comment (brtietz) [inline ssc/cmod_host_developer.cpp:2759]: I see where the parasitics are included in the real LCOE, but what about nominal?

**Review by brtietz — CHANGES_REQUESTED** [review]
My code review found a few typos worth fixing.

It also looks like lnte in Third Party uses CF_energy_net in the denominator - should this be updated as well?

**Comment by mjprilliman** [comment]
> If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate.
> 
> As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.
> 
> I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.

I'm cleaning up the cash flow outputs in the other FOM financial models (as well as fixing the issues Brian found). So it would now be possible to have the energy purchases and utility bill in the cash flow and Send to Excel for each. I'm not opposed to eliminating the option for some FOM configurations, but how do we decide which configs have that option and which do not?

**Comment by cpaulgilman** [comment]
> > If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate.
> > As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.
> > I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.
> 
> I'm cleaning up the cash flow outputs in the other FOM financial models (as well as fixing the issues Brian found). So it would now be possible to have the energy purchases and utility bill in the cash flow and Send to Excel for each. I'm not opposed to eliminating the option for some FOM configurations, but how do we decide which configs have that option and which do not?

I think we should make the option available for all FOM configurations. I'll commit my cashflow.lk that includes those outputs in the cash flow for all FOM configs so you can test with your changes to the financial models.

File: ssc/cmod_cashloan.cpp [hunk 4]
```diff
@@ -976,11 +1022,79 @@ class cm_cashloan : public compute_module
...
+                for (size_t m = 1; m <= 12; m++) {
+                    monthly_hourly_purchases = 0;
+                    monthly_hourly_fromgrid = 0;
+                    for (size_t d = 1; d <= util::days_in_month(int(m - 1)); d++) {
+                        for (size_t h = 0; h < 24; h++) { //monthly iteration for each year
+                            for (size_t n = 0; n < n_steps_per_hour; n++) {
+                                monthly_e_fromgrid[m-1] += year1_hourly_e_from_grid[util::hour_of_year(m, d, h) * n_steps_per_hour + n];
+                                monthly_gen_purchases[(a - 1) * 12 + m-1] += -gen_purchases[(size_t(a) - 1) * 8760 * n_steps_per_hour + n_steps_per_hour * util::hour_of_year(m, d, h) + n];
+                                if (year1_hourly_e_from_grid[h] != 0.0) {
+                                    //cf.at(CF_charging_cost_grid_month, a) += monthly_grid_to_batt[m] / (monthly_grid_to_batt[m] + monthly_grid_to_load[m]) * monthly_energy_charge[m] * charged_grid[a] / charged_grid[1] * cf.at(CF_util_escal_rate, a);
```
Comment (brtietz) [inline ssc/cmod_cashloan.cpp:1062]: Do you mind removing the commented out lines prior to the merge?

File: ssc/cmod_equpartflip.cpp [hunk 5]
```diff
@@ -1149,6 +1187,68 @@ class cm_equpartflip : public compute_module
...
+                for (size_t h = 0; h < 8760; h++) {
+                    cf.at(CF_energy_sales_value, i) += hourly_energy_calcs.hourly_sales()[h] * cf.at(CF_degradation, i) * ppa_value / 100.0 * ppa_multipliers[h];
+                    if (ppa_purchases) {
+                        cf.at(CF_energy_purchases_value, i) += -hourly_energy_calcs.hourly_purchases()[h] * cf.at(CF_degradation, i) * ppa_value / 100.0 * ppa_multipliers[h];
+                    }
+                }
+                if (!ppa_purchases) {
+                    cf.at(CF_energy_purchases_value, i) = 0.0;
+                    // Recompute this variable because the ppa_gen values (hourly_net) were all positve until now 
+                    //cf.at(CF_energy_net, i) = cf.at(CF_energy_sales, i) + cf.at(CF_energy_purchases, i); // Adding a positive and negative number
```
Comment (brtietz) [inline ssc/cmod_equpartflip.cpp:1247]: More commented out code - looks like the comment could also be removed?

File: ssc/cmod_equpartflip.cpp [hunk 6]
```diff
@@ -1059,62 +1067,92 @@ class cm_equpartflip : public compute_module
...
         }
 		
-
+        if (is_assigned("utility_bill_w_sys"))
+        {
+            size_t ub_count;
+            ssc_number_t* ub_arr;
+            ub_arr = as_array("utility_bill_w_sys", &ub_count);
+            if (ub_count != (size_t)(nyears + 1))
+                throw exec_error("singleowner", util::format("utility bill years (%d) not equal to analysis period years (%d).", (int)ub_count, nyears));
```
Comment (brtietz) [inline ssc/cmod_equpartflip.cpp:1076]: Issue with cmod name in exception

File: ssc/cmod_singleowner.cpp [hunk 7]
```diff
@@ -2979,27 +3008,28 @@ class cm_singleowner : public compute_module
...
+            - cf.at(CF_funding_equip1, i)
+            - cf.at(CF_funding_equip2, i)
+            - cf.at(CF_funding_equip3, i)
+            - cf.at(CF_funding_om, i)
+            - cf.at(CF_funding_receivables, i)
+            - cf.at(CF_funding_debtservice, i)
+            + cf.at(CF_reserve_interest, i)
+            - cf.at(CF_disbursement_debtservice, i) // note sign is negative for positive disbursement
+            - cf.at(CF_disbursement_om, i) // note sign is negative for positive disbursement
+            //+ cf.at(CF_energy_purchases_value, i);
```
Comment (brtietz) [inline ssc/cmod_singleowner.cpp:3032]: Why is this commented out for singleowner but not the other compute modules?

**Review by brtietz — COMMENTED** [review]
This is close. I think taking two steps for cleanliness would help a lot:

(1) Do a search for <//cf> and see if the line is supposed to be commented out. If yes, delete it and associated comments.

(2) Do a search for <throw exec_error("singleowner"> in compute modules that are not singleowner and update the cmod name appropriately.

The brackets in the above are just to indicate the search term without quotes (since the singleowner search ends with a quote)

**Review by brtietz — APPROVED** [review]
Great work! Thanks for doing the comment and exception cleanup. I think this is ready to go once the SAM side of things is ready to go.

**Pull request merged**